throw some exception when kubelet terminate csi node container

**What happened**:
{"level":"info","ts":1659103499.6545281,"msg":"Stopping and waiting for non leader election runnables"}
{"level":"info","ts":1659103529.6545928,"msg":"Stopping and waiting for leader election runnables"}
{"level":"info","ts":1659103529.6546292,"msg":"Stopping and waiting for caches"}
{"level":"info","ts":1659103529.654654,"msg":"Stopping and waiting for webhooks"}
{"level":"info","ts":1659103529.6546695,"msg":"Wait completed, proceeding to shutdown the manager"}
{"level":"error","ts":1659103529.654631,"logger":"setup","msg":"problem running manager","error":"failed waiting for all runnables to end within grace period of 30s: context deadline exceeded","stacktrace":"github.com/carina-io/carina/cmd/carina-node/run.subMain\n\t/workspace/github.com/carina-io/carina/cmd/carina-node/run/run.go:163\ngithub.com/carina-io/carina/cmd/carina-node/run.glob..func1\n\t/workspace/github.com/carina-io/carina/cmd/carina-node/run/root.go:48\ngithub.com/spf13/cobra.(*Command).execute\n\t/workspace/github.com/carina-io/carina/vendor/github.com/spf13/cobra/command.go:856\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/workspace/github.com/carina-io/carina/vendor/github.com/spf13/cobra/command.go:974\ngithub.com/spf13/cobra.(*Command).Execute\n\t/workspace/github.com/carina-io/carina/vendor/github.com/spf13/cobra/command.go:902\ngithub.com/carina-io/carina/cmd/carina-node/run.Execute\n\t/workspace/github.com/carina-io/carina/cmd/carina-node/run/root.go:55\nmain.main\n\t/workspace/github.com/carina-io/carina/cmd/carina-node/main.go:29\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:255"}
{"level":"info","ts":1659103529.654709,"logger":"controller.nodestorageresource","msg":"Shutdown signal received, waiting for all workers to finish","reconciler group":"carina.storage.io","reconciler kind":"NodeStorageResource"}
panic: close of closed channel

goroutine 1 [running]:
github.com/carina-io/carina/cmd/carina-node/run.subMain()
	/workspace/github.com/carina-io/carina/cmd/carina-node/run/run.go:165 +0xb88
github.com/carina-io/carina/cmd/carina-node/run.glob..func1(0x28b9440, {0x1920d0d, 0x3, 0x3})
	/workspace/github.com/carina-io/carina/cmd/carina-node/run/root.go:48 +0x1e
github.com/spf13/cobra.(*Command).execute(0x28b9440, {0xc000138050, 0x3, 0x3})
	/workspace/github.com/carina-io/carina/vendor/github.com/spf13/cobra/command.go:856 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0x28b9440)
	/workspace/github.com/carina-io/carina/vendor/github.com/spf13/cobra/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
	/workspace/github.com/carina-io/carina/vendor/github.com/spf13/cobra/command.go:902
github.com/carina-io/carina/cmd/carina-node/run.Execute()
	/workspace/github.com/carina-io/carina/cmd/carina-node/run/root.go:55 +0x25
main.main()
	/workspace/github.com/carina-io/carina/cmd/carina-node/main.go:29 +0x1c

**What you expected to happen**:
no any exception

**How to reproduce it**:
kubectl delete csinode-pod -n carina

--- a/cmd/carina-controller/run/node_metrics.go
+++ b/cmd/carina-controller/run/node_metrics.go
@@ -148,7 +148,6 @@ func (m *metricsExporter) Start(ctx context.Context) error {
 			return nil
 		}
 	}
-	return nil
 }
 
 // NeedLeaderElection implements controller-runtime's manager.LeaderElectionRunnable.

--- a/cmd/carina-controller/run/node_metrics.go
+++ b/cmd/carina-controller/run/node_metrics.go
@@ -127,19 +127,25 @@ func (m *metricsExporter) Start(ctx context.Context) error {
 		}
 	}()
 
-	ticker := time.Tick(1 * time.Minute)
-	for range ticker {
-		dm, err := vgMetrics()
-		if err == nil && len(dm) > 0 {
-			for _, m := range dm {
-				metricsCh <- m
+	ticker := time.NewTicker(1 * time.Minute)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			dm, err := vgMetrics()
+			if err == nil && len(dm) > 0 {
+				for _, m := range dm {
+					metricsCh <- m
+				}
 			}
-		}
-		vm, err := volumeMetrics()
-		if err == nil && len(vm) > 0 {
-			for _, v := range vm {
-				volumeCh <- v
+			vm, err := volumeMetrics()
+			if err == nil && len(vm) > 0 {
+				for _, v := range vm {
+					volumeCh <- v
+				}
 			}
+		case <-ctx.Done():
+			return nil
 		}
 	}
 	return nil

--- a/cmd/carina-controller/run/run.go
+++ b/cmd/carina-controller/run/run.go
@@ -31,8 +31,6 @@ import (
 	"github.com/carina-io/carina/utils"
 	"github.com/container-storage-interface/spec/lib/go/csi"
 	"google.golang.org/grpc"
-	corev1 "k8s.io/api/core/v1"
-	storagev1 "k8s.io/api/storage/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
@@ -107,26 +105,6 @@ func subMain() error {
 	}
 	if err := nodecontroller.SetupWithManager(mgr); err != nil {
 		setupLog.Error(err, "unable to create controller", "controller", "Node")
-		return err
-	}
-
-	// +kubebuilder:scaffold:builder
-
-	// pre-cache objects
-	if _, err := mgr.GetCache().GetInformer(ctx, &storagev1.StorageClass{}); err != nil {
-		return err
-	}
-	if _, err := mgr.GetCache().GetInformer(ctx, &corev1.Pod{}); err != nil {
-		return err
-	}
-	if _, err := mgr.GetCache().GetInformer(ctx, &corev1.PersistentVolume{}); err != nil {
-		return err
-	}
-	if _, err := mgr.GetCache().GetInformer(ctx, &carinav1.LogicVolume{}); err != nil {
-		return err
-	}
-
-	if _, err := mgr.GetCache().GetInformer(ctx, &corev1.Node{}); err != nil {
 		return err
 	}
 

--- a/cmd/carina-controller/run/run.go
+++ b/cmd/carina-controller/run/run.go
@@ -123,7 +123,6 @@ func subMain() error {
 	n := k8s.NewNodeService(mgr)
 
 	grpcServer := grpc.NewServer()
-	defer grpcServer.Stop()
 	csi.RegisterIdentityServer(grpcServer, driver.NewIdentityService())
 	csi.RegisterControllerServer(grpcServer, driver.NewControllerService(s, n))
 

--- a/cmd/carina-controller/run/run.go
+++ b/cmd/carina-controller/run/run.go
@@ -17,7 +17,6 @@
 package run
 
 import (
-	"context"
 	"fmt"
 	"net"
 
@@ -99,13 +98,12 @@ func subMain() error {
 	wh.Register("/pod/mutate", hook.PodMutator(mgr.GetClient(), dec))
 	//wh.Register("/pvc/mutate", hook.PVCMutator(mgr.GetClient(), dec))
 
-	stopChan := make(chan struct{})
-	defer close(stopChan)
+	ctx := ctrl.SetupSignalHandler()
 
 	// register controllers
 	nodecontroller := &controllers.NodeReconciler{
 		Client:   mgr.GetClient(),
-		StopChan: stopChan,
+		StopChan: ctx.Done(),
 	}
 	if err := nodecontroller.SetupWithManager(mgr); err != nil {
 		setupLog.Error(err, "unable to create controller", "controller", "Node")
@@ -115,7 +113,6 @@ func subMain() error {
 	// +kubebuilder:scaffold:builder
 
 	// pre-cache objects
-	ctx := context.Background()
 	if _, err := mgr.GetCache().GetInformer(ctx, &storagev1.StorageClass{}); err != nil {
 		return err
 	}
@@ -148,6 +145,7 @@ func subMain() error {
 	n := k8s.NewNodeService(mgr)
 
 	grpcServer := grpc.NewServer()
+	defer grpcServer.Stop()
 	csi.RegisterIdentityServer(grpcServer, driver.NewIdentityService())
 	csi.RegisterControllerServer(grpcServer, driver.NewControllerService(s, n))
 
@@ -159,11 +157,11 @@ func subMain() error {
 	}
 
 	// Http Server
-	e := newHttpServer(mgr.GetCache(), stopChan)
+	e := newHttpServer(mgr.GetCache(), ctx.Done())
 	go e.start()
 
 	setupLog.Info("starting manager")
-	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+	if err := mgr.Start(ctx); err != nil {
 		setupLog.Error(err, "problem running manager")
 		return err
 	}

--- a/cmd/carina-node/run/run.go
+++ b/cmd/carina-node/run/run.go
@@ -17,7 +17,6 @@
 package run
 
 import (
-	"context"
 	"errors"
 	"os"
 
@@ -31,7 +30,6 @@ import (
 	deviceManager "github.com/carina-io/carina/pkg/devicemanager"
 	"github.com/container-storage-interface/spec/lib/go/csi"
 	"google.golang.org/grpc"
-	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
@@ -71,23 +69,16 @@ func subMain() error {
 		return err
 	}
 
-	// pre-cache objects
-	ctx := context.Background()
-	if _, err := mgr.GetCache().GetInformer(ctx, &corev1.Pod{}); err != nil {
-		return err
-	}
+	ctx := ctrl.SetupSignalHandler()
 
 	// 初始化磁盘管理服务
-	stopChan := make(chan struct{})
-	defer close(stopChan)
-	dm := deviceManager.NewDeviceManager(nodeName, mgr.GetCache(), stopChan)
+	dm := deviceManager.NewDeviceManager(nodeName, mgr.GetCache(), ctx.Done())
 
 	podController := controllers.NewPodReconciler(
 		mgr.GetClient(),
 		nodeName,
 		dm.Partition,
 	)
-
 	if err := podController.SetupWithManager(mgr); err != nil {
 		setupLog.Error(err, "unable to create controller ", "controller", "podController")
 		return err
@@ -101,7 +92,6 @@ func subMain() error {
 		dm.VolumeManager,
 		dm.Partition,
 	)
-
 	if err := lvController.SetupWithManager(mgr); err != nil {
 		setupLog.Error(err, "unable to create controller", "controller", "LogicalVolume")
 		return err
@@ -110,14 +100,9 @@ func subMain() error {
 	nodeResourceController := controllers.NewNodeStorageResourceReconciler(
 		mgr.GetClient(),
 		nodeName,
-		stopChan,
+		ctx.Done(),
 		dm,
 	)
-
-	if _, err := mgr.GetCache().GetInformer(ctx, &corev1.Node{}); err != nil {
-		return err
-	}
-	// +kubebuilder:scaffold:builder
 
 	// Add metrics exporter to manager.
 	// Note that grpc.ClientConn can be shared with multiple stubs/services.
@@ -147,14 +132,13 @@ func subMain() error {
 	// 启动volume一致性检查
 	dm.VolumeConsistencyCheck()
 	// http server
-	e := newHttpServer(dm.VolumeManager, stopChan)
+	e := newHttpServer(dm.VolumeManager, ctx.Done())
 	go e.start()
 
 	go nodeResourceController.Run()
 	setupLog.Info("starting manager")
-	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+	if err := mgr.Start(ctx); err != nil {
 		setupLog.Error(err, "problem running manager")
-		close(stopChan)
 		return err
 	}
 

--- a/pkg/csidriver/runners/metrics_exporter.go
+++ b/pkg/csidriver/runners/metrics_exporter.go
@@ -157,7 +157,6 @@ func (m *metricsExporter) Start(ctx context.Context) error {
 			return nil
 		}
 	}
-	return nil
 }
 
 // NeedLeaderElection implements controller-runtime's manager.LeaderElectionRunnable.

--- a/pkg/csidriver/runners/metrics_exporter.go
+++ b/pkg/csidriver/runners/metrics_exporter.go
@@ -123,32 +123,38 @@ func (m *metricsExporter) Start(ctx context.Context) error {
 		}
 	}()
 
-	ticker := time.Tick(10 * time.Minute)
-	for range ticker {
-		vgList, err := m.volume.GetCurrentVgStruct()
-		if err == nil && len(vgList) > 0 {
-			for _, vg := range vgList {
-				metricsCh <- DeviceMetrics{
-					FreeBytes:   vg.VGFree,
-					TotalBytes:  vg.VGSize,
-					DeviceGroup: vg.VGName,
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			vgList, err := m.volume.GetCurrentVgStruct()
+			if err == nil && len(vgList) > 0 {
+				for _, vg := range vgList {
+					metricsCh <- DeviceMetrics{
+						FreeBytes:   vg.VGFree,
+						TotalBytes:  vg.VGSize,
+						DeviceGroup: vg.VGName,
+					}
 				}
 			}
-		}
 
-		volumeList, err := m.volume.VolumeList("", "")
+			volumeList, err := m.volume.VolumeList("", "")
 
-		if err == nil && len(volumeList) > 0 {
-			for _, v := range volumeList {
-				if !strings.HasPrefix(v.LVName, "volume") {
-					continue
-				}
-				volumeCh <- VolumeMetrics{
-					Volume:     v.LVName,
-					TotalBytes: v.LVSize,
-					UsedBytes:  float64(v.LVSize) * v.DataPercent / 100,
+			if err == nil && len(volumeList) > 0 {
+				for _, v := range volumeList {
+					if !strings.HasPrefix(v.LVName, "volume") {
+						continue
+					}
+					volumeCh <- VolumeMetrics{
+						Volume:     v.LVName,
+						TotalBytes: v.LVSize,
+						UsedBytes:  float64(v.LVSize) * v.DataPercent / 100,
+					}
 				}
 			}
+		case <-ctx.Done():
+			return nil
 		}
 	}
 	return nil

